File: Sources/Gantry/Services/PrinterInsights.swift
import Foundation

/// Local, vendor-neutral print history and maintenance counters. Printer firmware does not expose a
/// stable maintenance schedule across Bambu/Elegoo/Klipper/Prusa, so Gantry counts actual printing
/// time and lets each printer override the default intervals. HMS errors remain live printer alerts.
@MainActor
final class PrinterInsightsStore {
    static let shared = PrinterInsightsStore()
    static let didChange = Notification.Name("GantryPrinterInsightsDidChange")

    enum Result: String, Codable { case completed, failed, cancelled }
    enum Signal: Equatable { case none, planned, due(Int), urgent(Int) }

    struct HistoryEntry: Codable, Identifiable {
        let id: UUID
        let job: String
        let startedAt: Date
        let endedAt: Date
        let result: Result
        let durationSeconds: Double
    }

    struct TaskState: Codable {
        var intervalHours: Double
        var completedAtPrintHours: Double
        var snoozedUntil: Date?
    }

    struct PrinterRecord: Codable {
        var totalPrintSeconds: Double = 0
        var activeStartedAt: Date?
        var activeJob: String?
        var history: [HistoryEntry] = []
        var tasks: [String: TaskState] = [:]
    }

    struct TaskStatus: Identifiable {
        let id: String
        let title: String
        let intervalHours: Double
        let remainingHours: Double
        let overdueHours: Double
        let isDue: Bool
        let isUrgent: Bool
        let snoozedUntil: Date?
    }

    struct Snapshot {
        let totalPrintHours: Double
        let history: [HistoryEntry]
        let tasks: [TaskStatus]
        let consumedGrams: Double

        var completedCount: Int { history.filter { $0.result == .completed }.count }
        var unsuccessfulCount: Int { history.filter { $0.result != .completed }.count }
        var successPercent: Int? {
            guard !history.isEmpty else { return nil }
            return Int((Double(completedCount) / Double(history.count) * 100).rounded())
        }
    }

    struct TaskDefinition {
        let id: String
        let polish: String
        let english: String
        let defaultHours: Double
    }

    static let definitions: [TaskDefinition] = [
        .init(id: "clean-rods", polish: "Czyszczenie prowadnic", english: "Clean guide rods", defaultHours: 100),
        .init(id: "lubricate-axes", polish: "Smarowanie osi", english: "Lubricate axes", defaultHours: 200),
        .init(id: "inspect-belts", polish: "Kontrola pasków", english: "Inspect belts", defaultHours: 300),
        .init(id: "inspect-nozzle", polish: "Kontrola dyszy", english: "Inspect nozzle", defaultHours: 500)
    ]

    private let key = "printer-insights-v1"
    private var records: [String: PrinterRecord]
    private var lastPersistedAt = Date.distantPast

    private init() {
        if let data = BambuDefaults.shared.data(forKey: key),
           let decoded = try? JSONDecoder().decode([String: PrinterRecord].self, from: data) {
            records = decoded
        } else {
            records = [:]
        }
    }

    func observe(printer: SavedPrinter, previous: PrinterTelemetry?, current: PrinterTelemetry) {
        var record = records[printer.serial] ?? PrinterRecord()
        let now = current.lastUpdated ?? Date()
        let wasActive = previous?.state == .printing || previous?.state == .paused
        let isActive = current.state == .printing || current.state == .paused

        if isActive, record.activeStartedAt == nil {
            record.activeStartedAt = now
            record.activeJob = current.jobName
        }
        if current.state == .printing, let last = previous?.lastUpdated {
            // Protect the counter from a stale timestamp after sleep/reconnect; long sessions are
            // finalized from activeStartedAt when the print ends.
            let delta = max(0, min(now.timeIntervalSince(last), 120))
            record.totalPrintSeconds += delta
        }

        var result: Result?
        if current.state == .finished, previous?.state != .finished,
           wasActive || record.activeStartedAt != nil { result = .completed }
        else if wasActive, current.state == .error { result = .failed }
        else if wasActive, current.state == .idle, (previous?.progress ?? 0) > 0, (previous?.progress ?? 0) < 100 {
            result = .cancelled
        }
        if let result {
            let start = record.activeStartedAt ?? now
            let duration = max(0, now.timeIntervalSince(start))
            let job = current.jobName ?? previous?.jobName ?? record.activeJob ?? ""
            // One terminal transition may be repeated by partial telemetry; keep it idempotent.
            let duplicate = record.history.last.map {
                $0.result == result && $0.job == job && abs($0.endedAt.timeIntervalSince(now)) < 30
            } ?? false
            if !duplicate {
                record.history.append(.init(id: UUID(), job: job, startedAt: start, endedAt: now,
                                            result: result, durationSeconds: duration))
                if record.history.count > 100 { record.history.removeFirst(record.history.count - 100) }
            }
            record.activeStartedAt = nil
            record.activeJob = nil
        } else if !isActive, current.state != .offline, !wasActive {
            record.activeStartedAt = nil
            record.activeJob = nil
        }

        records[printer.serial] = record
        save(notify: result != nil, force: result != nil)
    }

    func snapshot(serial: String, polish: Bool) -> Snapshot {
        let record = records[serial] ?? PrinterRecord()
        let hours = record.totalPrintSeconds / 3600
        let tasks = Self.definitions.map { definition -> TaskStatus in
            let state = record.tasks[definition.id]
                ?? TaskState(intervalHours: definition.defaultHours, completedAtPrintHours: 0, snoozedUntil: nil)
            let remaining = state.completedAtPrintHours + state.intervalHours - hours
            let snoozed = state.snoozedUntil.map { $0 > Date() } ?? false
            let due = remaining <= 0 && !snoozed
            let overdue = max(0, -remaining)
            let urgent = due && overdue >= max(24, state.intervalHours * 0.15)
            return TaskStatus(id: definition.id, title: polish ? definition.polish : definition.english,
                              intervalHours: state.intervalHours, remainingHours: max(0, remaining),
                              overdueHours: overdue, isDue: due, isUrgent: urgent,
                              snoozedUntil: state.snoozedUntil)
        }
        let grams = SpoolbaseShared.spools.usageEvents
            .filter { $0.printerSerial == serial }
            .reduce(0) { $0 + $1.consumedGrams }
        return Snapshot(totalPrintHours: hours, history: Array(record.history.reversed()),
                        tasks: tasks, consumedGrams: grams)
    }

    func signal(serial: String, hmsCodes: [String]) -> Signal {
        let tasks = snapshot(serial: serial, polish: true).tasks
        let urgent = tasks.filter(\.isUrgent).count + (hmsCodes.isEmpty ? 0 : 1)
        if urgent > 0 { return .urgent(urgent) }
        let due = tasks.filter(\.isDue).count
        if due > 0 { return .due(due) }
        let now = Date()
        let planned = tasks.contains {
            !$0.isDue && !($0.snoozedUntil.map { $0 > now } ?? false)
                && $0.remainingHours <= max(24, $0.intervalHours * 0.1)
        }
        return planned ? .planned : .none
    }

    func complete(serial: String, taskID: String) {
        var record = records[serial] ?? PrinterRecord()
        let definition = Self.definitions.first { $0.id == taskID }
        var state = record.tasks[taskID]
            ?? TaskState(intervalHours: definition?.defaultHours ?? 200, completedAtPrintHours: 0, snoozedUntil: nil)
        state.completedAtPrintHours = record.totalPrintSeconds / 3600
        state.snoozedUntil = nil
        record.tasks[taskID] = state
        records[serial] = record
        save(notify: true, force: true)
    }

    func snooze(serial: String, taskID: String, days: Int = 7) {
        var record = records[serial] ?? PrinterRecord()
        let definition = Self.definitions.first { $0.id == taskID }
        var state = record.tasks[taskID]
            ?? TaskState(intervalHours: definition?.defaultHours ?? 200, completedAtPrintHours: 0, snoozedUntil: nil)
        state.snoozedUntil = Calendar.current.date(byAdding: .day, value: days, to: Date())
        record.tasks[taskID] = state
        records[serial] = record
        save(notify: true, force: true)
    }

    func setInterval(serial: String, taskID: String, hours: Double) {
        guard hours >= 1 else { return }
        var record = records[serial] ?? PrinterRecord()
        let definition = Self.definitions.first { $0.id == taskID }
        var state = record.tasks[taskID]
            ?? TaskState(intervalHours: definition?.defaultHours ?? hours, completedAtPrintHours: 0, snoozedUntil: nil)
        state.intervalHours = hours
        record.tasks[taskID] = state
        records[serial] = record
        save(notify: true, force: true)
    }

    private func save(notify: Bool, force: Bool = false) {
        let now = Date()
        if force || now.timeIntervalSince(lastPersistedAt) >= 60 {
            if let data = try? JSONEncoder().encode(records) { BambuDefaults.shared.set(data, forKey: key) }
            lastPersistedAt = now
        }
        if notify { NotificationCenter.default.post(name: Self.didChange, object: nil) }
    }
}

File: windows/Gantry.Windows/Services/PrinterInsights.cs
using System.Text.Json;
using Gantry.Models;

namespace Gantry.Services;

/// <summary>Vendor-neutral print history, statistics and per-printer maintenance schedules.</summary>
public static class PrinterInsights
{
    public enum PrintResult { Completed, Failed, Cancelled }
    public enum SignalKind { None, Planned, Due, Urgent }
    public sealed class HistoryEntry
    {
        public Guid Id { get; set; } = Guid.NewGuid();
        public string Job { get; set; } = "";
        public DateTime StartedAt { get; set; }
        public DateTime EndedAt { get; set; }
        public PrintResult Result { get; set; }
        public double DurationSeconds { get; set; }
    }
    public sealed class TaskState
    {
        public double IntervalHours { get; set; }
        public double CompletedAtPrintHours { get; set; }
        public DateTime? SnoozedUntil { get; set; }
    }
    public sealed class PrinterRecord
    {
        public double TotalPrintSeconds { get; set; }
        public DateTime? ActiveStartedAt { get; set; }
        public string? ActiveJob { get; set; }
        public List<HistoryEntry> History { get; set; } = new();
        public Dictionary<string, TaskState> Tasks { get; set; } = new();
    }
    public sealed record TaskStatus(string Id, string Title, double IntervalHours, double RemainingHours,
                                    double OverdueHours, bool IsDue, bool IsUrgent, DateTime? SnoozedUntil);
    public sealed record Snapshot(double TotalPrintHours, IReadOnlyList<HistoryEntry> History,
                                  IReadOnlyList<TaskStatus> Tasks, double ConsumedGrams)
    {
        public int CompletedCount => History.Count(value => value.Result == PrintResult.Completed);
        public int? SuccessPercent => History.Count == 0 ? null : (int)Math.Round(CompletedCount * 100.0 / History.Count);
    }
    public readonly record struct Signal(SignalKind Kind, int Count);
    private sealed record Definition(string Id, string Polish, string English, double Hours);

    private const string Key = "printer-insights-v1";
    private static readonly Definition[] Definitions =
    {
        new("clean-rods", "Czyszczenie prowadnic", "Clean guide rods", 100),
        new("lubricate-axes", "Smarowanie osi", "Lubricate axes", 200),
        new("inspect-belts", "Kontrola pasków", "Inspect belts", 300),
        new("inspect-nozzle", "Kontrola dyszy", "Inspect nozzle", 500),
    };
    private static readonly Dictionary<string, PrinterRecord> Records = Load();
    private static DateTime _lastPersistedAt = DateTime.MinValue;
    public static event Action? Changed;

    private static Dictionary<string, PrinterRecord> Load()
    {
        var json = Defaults.GetRaw(Key);
        if (string.IsNullOrWhiteSpace(json)) return new();
        try { return JsonSerializer.Deserialize<Dictionary<string, PrinterRecord>>(json) ?? new(); }
        catch { return new(); }
    }

    private static PrinterRecord Record(string serial)
    {
        if (!Records.TryGetValue(serial, out var value)) Records[serial] = value = new();
        return value;
    }

    public static void Observe(SavedPrinter printer, PrinterTelemetry? previous, PrinterTelemetry current)
    {
        var record = Record(printer.Serial);
        var now = current.LastUpdated ?? DateTime.Now;
        bool wasActive = previous?.State is PrinterState.Printing or PrinterState.Paused;
        bool active = current.State is PrinterState.Printing or PrinterState.Paused;
        if (active && record.ActiveStartedAt is null)
        {
            record.ActiveStartedAt = now;
            record.ActiveJob = current.JobName;
        }
        if (current.State == PrinterState.Printing && previous?.LastUpdated is { } last)
            record.TotalPrintSeconds += Math.Clamp((now - last).TotalSeconds, 0, 120);

        PrintResult? result = null;
        if (current.State == PrinterState.Finished && previous?.State != PrinterState.Finished &&
            (wasActive || record.ActiveStartedAt is not null)) result = PrintResult.Completed;
        else if (wasActive && current.State == PrinterState.Error) result = PrintResult.Failed;
        else if (wasActive && current.State == PrinterState.Idle && previous is { Progress: > 0 and < 100 }) result = PrintResult.Cancelled;
        if (result is { } terminal)
        {
            var start = record.ActiveStartedAt ?? now;
            var job = current.JobName ?? previous?.JobName ?? record.ActiveJob ?? "";
            bool duplicate = record.History.LastOrDefault() is { } lastEntry && lastEntry.Result == terminal &&
                             lastEntry.Job == job && Math.Abs((lastEntry.EndedAt - now).TotalSeconds) < 30;
            if (!duplicate)
            {
                record.History.Add(new HistoryEntry { Job = job, StartedAt = start, EndedAt = now,
                    Result = terminal, DurationSeconds = Math.Max(0, (now - start).TotalSeconds) });
                if (record.History.Count > 100) record.History.RemoveRange(0, record.History.Count - 100);
            }
            record.ActiveStartedAt = null; record.ActiveJob = null;
        }
        else if (!active && current.State != PrinterState.Offline && !wasActive)
        {
            record.ActiveStartedAt = null; record.ActiveJob = null;
        }
        Save(result is not null);
    }

    public static Snapshot GetSnapshot(string serial, bool polish)
    {
        var record = Record(serial);
        double hours = record.TotalPrintSeconds / 3600;
        var tasks = Definitions.Select(definition =>
        {
            if (!record.Tasks.TryGetValue(definition.Id, out var state))
                state = new TaskState { IntervalHours = definition.Hours };
            double remaining = state.CompletedAtPrintHours + state.IntervalHours - hours;
            bool due = remaining <= 0 && !(state.SnoozedUntil is { } until && until > DateTime.Now);
            double overdue = Math.Max(0, -remaining);
            return new TaskStatus(definition.Id, polish ? definition.Polish : definition.English,
                state.IntervalHours, Math.Max(0, remaining), overdue, due,
                due && overdue >= Math.Max(24, state.IntervalHours * .15), state.SnoozedUntil);
        }).ToList();
        double grams = SpoolbaseShared.Spools.UsageEvents.Where(value => value.PrinterSerial == serial).Sum(value => value.ConsumedGrams);
        return new Snapshot(hours, record.History.AsEnumerable().Reverse().ToList(), tasks, grams);
    }

    public static Signal GetSignal(string serial)
    {
        var tasks = GetSnapshot(serial, true).Tasks;
        int urgent = tasks.Count(value => value.IsUrgent);
        if (urgent > 0) return new(SignalKind.Urgent, urgent);
        int due = tasks.Count(value => value.IsDue);
        if (due > 0) return new(SignalKind.Due, due);
        bool planned = tasks.Any(value => !value.IsDue && !(value.SnoozedUntil is { } until && until > DateTime.Now)
                                              && value.RemainingHours <= Math.Max(24, value.IntervalHours * .1));
        return planned ? new(SignalKind.Planned, 0) : new(SignalKind.None, 0);
    }

    public static void Complete(string serial, string taskId) => Change(serial, taskId, completed: true);
    public static void Snooze(string serial, string taskId) => Change(serial, taskId, snooze: true);
    public static void SetInterval(string serial, string taskId, double hours)
    {
        if (hours >= 1) Change(serial, taskId, interval: hours);
    }

    private static void Change(string serial, string taskId, bool completed = false, bool snooze = false, double? interval = null)
    {
        var record = Record(serial);
        if (!record.Tasks.TryGetValue(taskId, out var state))
        {
            var definition = Definitions.FirstOrDefault(value => value.Id == taskId);
            record.Tasks[taskId] = state = new TaskState { IntervalHours = definition?.Hours ?? 200 };
        }
        if (completed) { state.CompletedAtPrintHours = record.TotalPrintSeconds / 3600; state.SnoozedUntil = null; }
        if (snooze) state.SnoozedUntil = DateTime.Now.AddDays(7);
        if (interval is { } value) state.IntervalHours = value;
        Save(true);
    }

    private static void Save(bool notify)
    {
        if (notify || (DateTime.Now - _lastPersistedAt).TotalSeconds >= 60)
        {
            Defaults.SetRaw(Key, JsonSerializer.Serialize(Records));
            _lastPersistedAt = DateTime.Now;
        }
        if (notify) Changed?.Invoke();
    }
}
